sph/include/sph/hydro_ve/eos_gpu.cu

```diff
@@ -59,22 +59,46 @@ __global__ void cudaEOS(size_t firstParticle, size_t lastParticle, Tm mui, Tt ga
 }
 
 template<class Tt, class Tm, class Thydro>
-void computeEOS(size_t firstParticle, size_t lastParticle, Tm mui, Tt gamma, const Tt* temp, const Tm* m,
+__global__ void cudaEOS_u(size_t firstParticle, size_t lastParticle, Tm mui, Tt gamma, const Tt* u, const Tm* m,
+                          const Thydro* kx, const Thydro* xm, const Thydro* gradh, Thydro* prho, Thydro* c, Thydro* rho,
+                          Thydro* p)
+{
+    unsigned i = firstParticle + blockDim.x * blockIdx.x + threadIdx.x;
+    if (i >= lastParticle) return;
+
+    Thydro p_i;
+    Thydro rho_i         = kx[i] * m[i] / xm[i];
+    util::tie(p_i, c[i]) = idealGasEOS_u(u[i], rho_i, gamma);
```

Comment:
I would put the if `u == nullptr` here and avoid two separate CUDA kernels. Performance-wise the difference is practically zero.

Reply:
Changed.